### PR TITLE
avoid recursive expansion of env variables

### PR DESCRIPTION
**Pull request by elahehrashedi**
bug fix: https://github.com/microsoft/vscode-cmake-tools/issues/2501

### CODE REVIEW

File: src/expand.ts [hunk 1]
```diff
@@ -169,7 +169,14 @@ async function expandStringHelper(tmpl: string, opts: ExpansionOptions) {
...
         const full = mat[0];
         const varname = mat[1];
-        const repl = fixPaths(env[varname]) || '';
+        const repl: string = fixPaths(env[varname]) || '';
+        // Avoid replacing an env variable by itself, e.g. PATH:env{PATH}.
+        const env_re4 = RegExp(`\\$env\\{(${varValueRegexp})\\}`, "g");
+        mat = env_re4.exec(repl);
+        const varnameRepl = mat ? mat[1] : undefined;
+        if (varnameRepl && varnameRepl === varname) {
+            break;
```
Comment (bobbrow) [inline src/expand.ts:178]: I think we just want to log here.  We already have a failsafe that stops the recursion, just make sure we don't log the same warning 10 times.

**Review by bobbrow — CHANGES_REQUESTED** [review]

File: src/expand.ts (hunk 1 above)
Comment (elahehrashedi) [reply src/expand.ts:178]: @bobbrow we try to expand this 3 times, in `expandConfigurePreset` and `expandBuildPreset`, so although we won't see the error 10 times, but we will see it 3 times. IS that fien?

Comment (bobbrow) [reply src/expand.ts:184]: That's ok.

File: src/expand.ts [hunk 2]
```diff
@@ -103,17 +103,21 @@ export async function expandString<T>(tmpl: string | T, opts: ExpansionOptions):
...
         result = expansion.result;
         didReplacement = expansion.didReplacement;
+        circularReference = expansion.circularReference;
         i++;
-    } while (i < MAX_RECURSION && opts.recursive && didReplacement);
+    } while (i < MAX_RECURSION && opts.recursive && didReplacement && !circularReference);
 
-    if (i === MAX_RECURSION) {
+    if (circularReference) {
+        log.warning(localize('circular.variable.reference', 'Circular variable reference found: {0}', circularReference));
```
Comment (bobbrow) [inline src/expand.ts:119]: This gives the string with the reference, like "${PATH}", but not the variable name that went with it.  Can we save `full` and `varname` down by line 183?

**Review by bobbrow — CHANGES_REQUESTED** [review]

**Review by bobbrow — APPROVED** [review]

**Pull request merged**